KVStore & derived classes: design docs & implementation

davidsaada opened the pull request:
KVStore & derived classes: design docs & implementation

@dannybenor @yossi2le @offirko @theamirocohen  Please review changes

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+## Global Key Value interface
+
+A parallel key-value API is provided as global C-style functions. This API performs a limited type of mapping of partition or mount point names present in the keys. For each of the APIs defined in KVStore, the global version will extract a partition prefix from the key name. The prefix must be in the form "/partition/key-name". Then a lookup will be performed to map the partition name to a concre ...
+In the case of iteration APIs, the pattern must include the partition (in the form of "/partition/<pattern>").  
```
fix pattern

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+        ROLLBACK_PROTECT_FLAG = (1 << 3),
+    };
+ 
+    static const uint32_t MAX_KEY_LENGTH = 128;
```
Change also in Confluence to 128

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+## Design basics
+
+KVStore defines a simple, key value store like API set via this interface class. Classes implementing this interface store pairs of keys and values, where the keys are represented as character strings and the values are represented as binary blobs. Core APIs here are *get* and *set*, providing read and write access by key to the value in a single call. *remove* completes the set of core APIs. Thi ...
+APIs also support an "incremental set" mode, allowing the implementing class to aggregate chunks of data for the set operation. This is for the case the caller needs to generate large portions of data, but doesn't wish to allocate large buffers for a single set operation.  
```
Fix the explanation of incremental read and add explanation for incremental read

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size);
```
add offset to get

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+int kv_get_info(const char *full_name_key, kv_info_t *info);
+int kv_remove(const char *full_name_key);
+ 
+// Incremental set API
```
Remove incremental API

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
````diff
@@ -0,0 +1,456 @@
+int kv_iterator_close(kv_iterator_t it);
+```
+
+## Attachment and lookup APIs
````
Mapping API

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+
+## Attachment and lookup APIs
+
+In order to use the global C style APIs, we need APIs to attach the partition name to the instance of the implementing KVStore class, typically called once at initialization time. So, for example if we have the `TDBStore` class implementing KVStore, whose instance is `tdbs` and we want to map it to the name "tdbstore", so each full name will have "/tdbstore/" preceding the key name, we will attac ...
```
Refrace to explain the mapping of prefix to a instance

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+## Attachment and lookup APIs
+
+In order to use the global C style APIs, we need APIs to attach the partition name to the instance of the implementing KVStore class, typically called once at initialization time. So, for example if we have the `TDBStore` class implementing KVStore, whose instance is `tdbs` and we want to map it to the name "tdbstore", so each full name will have "/tdbstore/" preceding the key name, we will attac ...
+These APIs will be part of a different header file ("kv_attach.h"), as they serve the integration code and not the KVStore user code.  
```
kv_map.h?

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+// Attach and detach
+int kv_init();
+int kv_attach(const char *partition_name, KVStore *kv_instance);
+int kv_detach(const char *partition_name, KVStore *kv_instance);
```
remove kv_instance from kv_detach

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+int res;
+
+// Initialize TDBStore
+res = tdbstore.init();
```
Please change tdbstore into kvstore in the example

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+// One can use SlicingBlockDevice if we want a partition.
+SPIFBlockDevice bd(PTE2, PTE4, PTE1, PTE5);
+
+// Instantiate tdbstore with our block device
```
Add the meaning of 64 to this comment

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+char buf[8];
+
+KVSTore::set_handle_t handle;
+res = tdbstore.set_start(&handle, "Key2", data_size, 0);
```
Check if we need the & in &handle

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+res = tdbstore.init();
+
+// Add "Key1"
+res = tdbstore.set("Key1", "Value of key 1", 12, 0);
```
Use sizeof instead of literals

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+int res;
+
+// Attachment code. Should be called at initialization
+res = kv_init();
```
Add title "Assuming this code exists somewhere...."

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+
+kv_set_handle_t handle;
+res = kv_set_start(&handle, "/tdbstore/Key2", data_size, 0);
+for (int i = 0; i < data_size / sizeof(buf); i++) {
```
Remove example of incremental (non-existing) API

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,456 @@
+// Iterate over all keys starting with "Key" in tdbstore
+res = 0;
+kv_iterator_t it;
+kv_iterator_open(&it, "/tdbstore/Key*");
```
Remove * from terator_open and update the name of the parameter

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instance is instantiated and initialized. 
```
We actually need 2 kvstore instances, one for the data and one for RBP

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+![SecureStore Layers](./SecureStore_layers.jpg)
```
Add an additional kvstore to the drawing

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+
+As such, it offers all KVStore APIs. Stored data is protected with the following security options (which can be selected using the creation flags at set):  
+
+- Encryption: Data is encrypted using the AES-CTR encryption method. Nonce is randomly generated and each encryption chunk (16 bytes) uses the incrementing counter.   
```
Nonce is randomly --> IV is composed from a 8 bytes nonce randomly generated and 8 bytes counter
Key is derived from devicekey (https://github.com/ARMmbed/mbed-os/tree/master/features/device_key) using the unique ID+"ENC" as salt.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+As such, it offers all KVStore APIs. Stored data is protected with the following security options (which can be selected using the creation flags at set):  
+
+- Encryption: Data is encrypted using the AES-CTR encryption method. Nonce is randomly generated and each encryption chunk (16 bytes) uses the incrementing counter.   
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data.
```
Key is derived from devicekey (https://github.com/ARMmbed/mbed-os/tree/master/features/device_key) using the unique ID+"AUTH" as salt.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce and external RBP key. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
+
+![SecureStore Record](./SecureStore_record.jpg)
```
Please add an explanation of the different fields

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+- Encryption: Data is encrypted using the AES-CTR encryption method. Nonce is randomly generated and each encryption chunk (16 bytes) uses the incrementing counter.   
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data.
+- Rollback protection: (Requires authentication) CMAC is stored in a designated rollback protected storage (also of KVStore type) when set and compared with the stored one at get.
+- Write once: Key can only be stored once and can't be removed. 
```
Write once should be also a feature of TDBStore. CMAC of write once should also be written as write once in the RBP kvstore

davidsaada replied:
@dannybenor What about FileSystemStore?

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP Keys are generated using a 32-bit incremental counter, which is incremented whenever we have a new key. Start value of the counter is achieved by iterating over the rollback protection KVStore (at initialization) and getting the maximal value. 
```
start value is a constant when the securestore is empty and later on by iterating ...

davidsaada replied:
This is true if use NVStore. If RBP is a KVStore, we will use a string, which can have a prefix (like "RBP12").

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+    Mutex _mutex;
+    KVStore *_underlying_kv;
+    KVStore *_rbp_kv;
+    void *_entropy;
```
is void * the right type in mbedtls?

davidsaada replied:
Yes - casted in CPP file to avoid exposure of type in header (a known C++ problem).

yossi2le commented on docs/design-documents/features/storage/TDBStore/TDBStore_design.md:
```diff
@@ -0,0 +1,461 @@
+GC is invoked in the following cases:
+1. When the active area is exhausted.
+2. During initialization, when a corruption is found while scanning the active area. In this case, GC is performed up to the record preceding the corruption.
+
```
GC is expensive and in an internal flash, it also blocking interrupts. Therefore I think it can be a good idea disabling GC when it founds out the area is full and there is nothing more to clean.
This will elimanate any further retries till a remove function is called.

davidsaada replied:
After discussing it with @dannybenor we decided to leave it as is. This is a very extreme case, and returning an error should be sufficient to cover it.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
+- Call `_rbp_kv` `get` API with on a local `read_cmac` variable, key `ext_rbp_key`, size 16
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
```
We need a different error like "RBP Error"

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,430 @@
+- Fill fields in `info` according to `metadata`
+- Return OK   
+
+**remove function**
```
Please fix with the right pseudo code

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
```
Should _mutex also protect this if check?
If so, I think _mutex should be created at constructor or statically otherwise it wouldn't be able to protect init().

davidsaada replied:
Well, pseudo code doesn't show all code, but basic concepts. May put this initialization in the constructor as you suggested, or use an atomic counter as done in NVStore.

dannybenor replied:
One big value of pseudo code will be the security review. I think this should be the criteria to add or remove functionality from the pseudo code

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+Pseudo code:  
+- if `_is_initialized` return OK
+- Create and take `_mutex`
+- Create the FileSystemStore directory if not existing
```
typo : "if doesn't exist"

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+`virtual int deinit();`
+
+Pseudo code:  
+- if not `_is_initialized` return OK
```
_mutex should be taken prior to this check

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+- if `_is_initialized` return OK
+- Create and take `_mutex`
+- Create the FileSystemStore directory if not existing
+- Using the FileSystem `dir_size` API, count number of files and store in `_num_keys` 
```
Do we need to authenticate the files first?

davidsaada replied:
No. Authentication is only performed in SecureStore layer. We perform an error checking in TDBStore as keys are sequential (so first error influences all following keys), but here it's not needed.

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+- if not `_is_initialized` return error
+- Find file `key` under the FileSystemStore directory. If not existing, return not found error
+- Take `_mutex` 
+- Using the FileSystem `stat` API, extract file size, set `actual_size` as file size - metadata size
```
Earlier it said that 'get' is implemented "using open-read-close"
Do we want to mention it here?

davidsaada replied:
Thanks, now specifically mentioned (also in writes).

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+- Find file `key` under the FileSystemStore directory. If not existing, return not found error
+- Take `_mutex` 
+- Using the FileSystem `stat` API, extract file size, set `actual_size` as file size - metadata size
+- Using `file_seek` API, seek to `offset`
```
Will offset always be 0? 
if so, I think fopen will get you to offset 0.

davidsaada replied:
It won't always be zero, but I'll add the seek only on the case it's non zero.

offirko commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,366 @@
+- Using `file_seek` API, seek to `offset`
+- Read data (excluding metadata size) to `buffer`. Size is minimum of buffer size and remainder of data
+- Read remainder into a `key_metadata_t` structure
+- If data size doesn't match size in metadata structure or wrong magic, return a data corrupt error. 
```
I think we need some elaboration about 'magic' here (even if its mentioned in other documents)

davidsaada replied:
Adding explanations on all fields.

davidsaada commented:
@dannybenor @yossi2le  @offirko @theamirocohen 
Docs modified following your comments. Please re-review.
Quick links: [KVStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/KVStore/KVStore_design.md) [TDBStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/TDBStore/TDBStore_design.md)  [FileSystemStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md) [SecureStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/SecureStore/SecureStore_design.md)

Commit:
KVStore & derived classes: design docs & implementation

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,462 @@
+`int kv_attach(const char *partition_name, KVStore *kv_instance);`
+
+Pseudo code:  
+- Duplicate `partition_name` and `kv_instance` to last entry in `kvstore_attachment_table`
```
Duplicate? Seems like a mistake

davidsaada replied:
Well, I could have used the term "copy", but as partition_name is a character string, the term duplicate seems more appropriate.

dannybenor commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
````diff
@@ -0,0 +1,462 @@
+
+```C
+// Assume TDBtore is already instantiated and initialized
+extern TDBStore tdbstore;
````
Why extern?

davidsaada replied:
Meaning here was that TDBStore exists somewhere outside this code. Otherwise, I would have needed to show how to instantiate and initialize it.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+Encryption, authentication & rollback protection features use an integer "unique key ID". This ID is incremented whenever we have a new key. Its initial value is determined at init time by iterating over all RBP keys in the rollback protection storage, and taking the maximal value + 1.    
+
+
+![SecureStore Layers](./SecureStore_layers.jpg)
```
Can you add the word "OR" on top of the dotted line that separates TDBStore from FileSystemStore?

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce and unique key. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
```
unique key --> unique key ID to avoid confusion with the encryption/authentication key

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+
+![SecureStore Record](./SecureStore_record.jpg)
+
+Fields are:
```
What about a version or entry type field? We may need if we will change the format of the header for backward compatibility

davidsaada replied:
Ugh... Rev. became reserved. Will change.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+Pseudo code:  
+- if `_is_initialized` return OK
+- Take `_mutex`
+- Initialize `_entropy` with TLS entropy APIs
```
Maybe better to do lazy initialization of entropy. This is an expensive activity, and better to do later than is startup, after entropy is gathered. Also for read-only you will not need entropy.

davidsaada replied:
This means I'll have to ask whether entropy is initialized in each set function. Sounds a bit awkward to me.

dannybenor replied:
Entropy initialization looks innocent but it can be heavy duty

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+`virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);`
+
+Pseudo code:  
+- Call `set_start` with all fields and a local `set_handle_t` variable
```
RBP logic is missing here. Important for the security review. Get received the right attention :)

davidsaada replied:
RBP logic is handled in set_start and set_finalize. Set function just calls set_start, set_add_data and set_finalize in order to prevent duplication of code.

dannybenor replied:
ok. I missed that you are calling set_start in SecureStore and not in tdbstore

davidsaada replied:
I do it in both cases.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+`virtual int get_info(const char *key, info_t *info);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
```
Maybe good to clarify that this function returns the size of the original data passed by the caller to set

davidsaada replied:
No - it returns a success/failure return code (like all others do).

dannybenor replied:
I mean returns in the info the size of the original data

davidsaada replied:
Not sure I understand - the caller passes an `info_t` structure. The function just fills it up.

offirko replied:
@dannybenor - Does the info consist just of size?
Do we need to authenticate in this function?

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+
+Pseudo code:
+
+- If `offset` + `data_size` > data size in handle, return error
```
You did not mention offset before. Do you mean actual data exceeds final_data_size passed in set_start?

davidsaada replied:
Offset is stored in the handle (maybe I should be more elaborate), so yes - this checks that I don't exceed final_data_size.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+extern TDBStore tdbstore;
+
+// Instantiate SecureStore with tdbstore as underlying key value store
+SecureStore secure_store(&tdbstore);
```
What about rbp_tdbstore for rollback protection? it is optional but how do you pass it?

davidsaada replied:
You are right, will add.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+ 
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is "ENC" + unique key ID (see below). 
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data. The key used for generating the CMAC is is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where sal ...
```
typo: "is is"

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
```
why 'value_data' and not just 'data'?
will data_size be the scratch_buf size 32?

davidsaada replied:
Value data was decided in confluence as a term to tell it's the data of the *value* (in the Key Value store).
Regarding data size - Yes. If a SecureStore `set` function calls a TDBStore `set_add_data`, the it will be 32 (or less).

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+- Initialize `_entropy` with TLS entropy APIs
+- Using `DeviceKey` APIs, get the device key
+- Allocate `_scratch_buf` as a 32 byte array
+- Iterate over all keys is `_rbp_kv` and set `_unique_key_id_counter` to the maximal key + 1   
```
typo: 'all keys in'

davidsaada replied:
Was removed anyway.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+
+Pseudo code:  
+- Call `set_start` with all fields and a local `set_handle_t` variable
+- Call `set_add_data` with `buffer` and `size`
```
is 'set_add_data' called with 'size' argument or with _scratch_buf size?

davidsaada replied:
Here it is called with `size`. I only use scratch buf for generating my own data (like encrypted chunks). Here I call it with the user data.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+- Set `current_offset` to 0
+- While `data_size` > 0
+	- If `current_offset` between `offset` and `actual_size`
+		- Set `dest_buf` to `buffer` and `chunk_size` to `actual_size`
```
is the decrypted data being copied to user buffer before authentication was completed?

davidsaada replied:
See answer below.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+	- Decrement `data_size` by `chunk_size`
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- Generate CMAC on local `cmac` variable 
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
```
if the data was already copied to user buffer, do we erase it in case of "data corrupt error" ?

davidsaada replied:
Very good catch. Will change it.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+	- If key exists, return "already exists" error
+- if flags include rollback protection flag
+	- Call `_underlying_kv` `get` API with `metadata` size into `metadata` field (to fill `unique_key_id` field)
+	- If key doesn't exist, incremet `_unique_key_id_counter` and fill it in `metadata` field
```
if 'key' doesnt exist what will return in 'metadata'? will we be able to fill it?

davidsaada replied:
It will have the value the user filled in originally. 
I didn't write it, but if `_underlying_kv` `get` API failes, it will return an error. Errors should be checked throughout the code and handled properly. I just didn't want to write it everywhere.

offirko commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,444 @@
+
+Pseudo code:
+- Take `_mutex`  
+- Call `_underlying_kv` `iterator_next` with `underlying_it` field
```
_underlying_kv might contain un-authenticated data - can it pose any risk?

davidsaada replied:
I guess not. We also don't encrypt the name, so I think it's the data that matters (and get will fail if we try to read the data with this key). @dannybenor What do you think?

dannybenor replied:
@offirko @davidsaada Names are not secret, therefore not encrypted.
The existence of a file can have security implications, and it became complex... For example, "provisioning done".
I think we can limit the security checks to get and get_info and document that iterator are not authenticated

davidsaada commented:
Docs updated. Main changes:
1. Added reserved area to TDBStore (for storage of ROT)
2. Changed FileSystemStore APIs to use retargeting APIs
3. SecureStore: Use names for underlying RBP store instead of numbers

davidsaada commented:
@dannybenor @yossi2le @offirko @theamirocohen

davidsaada commented:
Quick links once again: [KVStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/KVStore/KVStore_design.md) [TDBStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/TDBStore/TDBStore_design.md)  [FileSystemStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md) [SecureStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/SecureStore/SecureStore_design.md)

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,441 @@
+- Data size: Size of user data
+- Flags: User flags
+- Nonce: Nonce field used in IV
+- Pad: Pad data to a multiple of 16 bytes (due to encryption)
```
What is the relation between encryption and 16 bytes padding? Seems more like a minimal write size than encryption

davidsaada replied:
That's the SecureStore doc, so write size is not relevant (it's only relevant to TDBStore). I was thinking that a 16 byte padding would make decryption work easier, as decryption is performed in a 16 byte chunks. Without this padding, I would need to read the tail to a separate buffer in order to decrypt it.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,441 @@
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP Keys are generated using a 32-bit incremental counter, which is incremented whenever we have a new key. Start value of the counter is achieved by iterating over the rollback protection KVStore (at initialization) and getting the maximal value. 
```
Remove the 32 bit incremental counter

dannybenor commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,377 @@
+
+Pseudo code:  
+- if not `_is_initialized` return "not initialized" error
+- Find file `key` under the FileSystemStore directory. If not existing, return "not found" error
```
Why not try to open for read instead of "find"? If error then return error...
@offirko

davidsaada replied:
Fixed.

dannybenor commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,377 @@
+
+Pseudo code:  
+- if not `_is_initialized` return "not initialized" error
+- Find file `key` under the FileSystemStore directory. If not existing, return "not found" error.
```
open for read instead of find?

davidsaada replied:
Fixed.

dannybenor commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,377 @@
+private:
+    Mutex _mutex;
+    size_t *_max_keys;
+    size_t *_num_keys;
```
What is the usage of _num_keys? There is no mention in pseudo code
@offirko

davidsaada replied:
You are right. It is used for checking whether we exceed max keys. Will add in set API.
And if you mentioned it, this is one of the open issues - whether to have a similar limitation on number of keys as TDBStore does, while here there's no real need for it.

dannybenor commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,441 @@
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
```
Regarding get and get_info functions: If record is not found we need to check also if they exist in rbp_kvstore

davidsaada replied:
And do what if we find them there?

dannybenor replied:
return authentication error

davidsaada replied:
OK. Do we remove the key in RBP storage?
Edit: Reason I ask is that if we decide to remove it, then it makes sense to perform this check. If not, then it's just a difference between a "Not found" error and authentication error.

dannybenor replied:
We  do not remove. We need the hint in RBP that at some time the file existed in normal KVSTore

davidsaada commented:
@dannybenor @offirko Pushed an incremental commit based on your comments.
